cache-explainer.md

```diff
@@ -0,0 +1,104 @@
+
+## Authors
+
+WebML Working Group participants
+
+## Participate
+
+- https://github.com/webmachinelearning/webnn/issues/807
+
+## Table of contents
+
+1. [Introduction](#introduction)
+1. [Goals](#goals)
+1. [Non-goals](#non-goals)
+1. [User research](#user-research)
+1. [Use cases](#use-cases)
+1. [Proposed API](#proposed-api)
+1. [Considered alternatives](#considered-alternatives)
+1. [Related work](#related-work)
+1. [Privacy and security considerations](#privacy-and-security-considerations)
+1. [References](#references)
+
+## Introduction
+
+The WebNN API enables web applications to perform ML model inference by constructing a graph representation of the model ([`MLGraphBuilder`](https://www.w3.org/TR/webnn/#mlgraphbuilder)), compiling it into a native format ([`MLGraph`](https://www.w3.org/TR/webnn/#mlgraph)), and executing it via [`MLContext.dispatch()`](https://www.w3.org/TR/webnn/#api-mlcontext-dispatch). However, compiling large models on certain devices, such as NPUs, can be time-consuming. To address this, we propose an explicit API for caching compiled graphs, allowing web applications to save and reuse them, thereby reducing the overhead of repeated compilation.
+
+This proposal documents ongoing discussions in the W3C WebML Working Group and builds on existing mechanisms in frameworks like ONNX Runtime.
+
+## Goals
+
+- Provide a mechanism for web applications to save and load compiled `MLGraph` objects.
+- Reduce the time required for repeated ML model inference by avoiding redundant graph compilation.
+- Ensure compatibility with existing WebNN API constructs and workflows.
+
+## Non-goals
+
+- This proposal does not aim to define a universal format for graph serialization across all frameworks.
+- It does not address caching mechanisms for non-WebNN APIs or other types of computational graphs.
+- Cross-origin model sharing is out of scope.
+
+## User research
+
+[If any user research has been conducted to inform the design choices presented,
+discuss the process and findings.
+We strongly encourage that API designers consider conducting user research to
+verify that their designs meet user needs and iterate on them,
+though we understand this is not always feasible.]
+
+## Use cases
+
+### Reduce time to first inference
+
+A web application performing real-time image recognition can save the compiled graph after the first inference. If the page is reloaded, subsequent inferences reuse the cached graph, significantly reducing latency by avoiding both the model redownload and recompilation steps.
+
+## Proposed API
+
+```webidl
+partial interface MLContext {
+  Promise<sequence<DOMString>> listGraphs();
+  Promise<MLGraph> loadGraph(DOMString key);
+  Promise<undefined> saveGraph(DOMString key, MLGraph graph);
+  undefined deleteGraph(DOMString key);
+};
+```
+
+1. **`listGraphs()`**: Returns a list of keys for all cached graphs.
+2. **`loadGraph(key)`**: Loads a cached graph associated with the given key.
```

Comment:
Note that a graph may be evicted from the cache due to storage pressure or browser/platform updates which render previously compiled graphs invalid. Developers should consider the level of durability to be somewhere between IndexedDB and the HTTP cache. I'm not sure if there is a specific concept from the storage specs we could reference here.

Reply:
Added a "A note on persistence" [sic] https://github.com/webmachinelearning/webnn/pull/862/commits/20ed41475d6cd99ce793ebf76d19baa7275e419c